File: src/ForceSignal.Domain/Rules/RulesProfile.cs
using System.Collections.Immutable;

namespace ForceSignal.Domain.Rules;

/// <summary>
/// How a hull's damage track is divided into rows, which decides how many threshold checks a ship
/// faces on its way to being destroyed.
/// </summary>
public enum ThresholdRowMode
{
    /// <summary>The same number of rows for every hull, whatever its size.</summary>
    FixedRows,

    /// <summary>Rows by size band, so a smaller hull faces fewer checks before it dies.</summary>
    ByShipClass
}

/// <summary>What one beam die scores against one level of screening.</summary>
/// <param name="DieFace">The face rolled.</param>
/// <param name="ScreenLevel">The target's screen level.</param>
/// <param name="Damage">Points that face scores through that screening.</param>
public sealed record BeamDamageEntry(int DieFace, int ScreenLevel, int Damage);

/// <summary>What one point-defence die shoots down.</summary>
/// <param name="DieFace">The face rolled.</param>
/// <param name="Kills">How many incoming it accounts for.</param>
public sealed record PointDefenseEntry(int DieFace, int Kills);

/// <summary>What a landed fighter group's turnaround roll means for it.</summary>
/// <param name="DieFace">The face rolled.</param>
/// <param name="IsGroundedForGame">True when the group does not fly again this game.</param>
/// <param name="TurnsBeforeRelaunch">Turns the deck crews need before it can go out again.</param>
public sealed record TurnaroundEntry(int DieFace, bool IsGroundedForGame, int TurnsBeforeRelaunch);

/// <summary>
/// Every number a Full Thrust match is played against, gathered in one place and supplied by the
/// player rather than shipped with this app.
/// </summary>
/// <remarks>
/// This type used to carry two ready-made profiles with the published numbers written into them,
/// and the rest of the numbers were scattered through the rule classes as constants. Both were the
/// same mistake: the engine owns the <em>procedures</em> - what is rolled against what, what order
/// things happen in, what a screen does to a die - and the player owns every <em>number</em> those
/// procedures read, off their own rulebook and record cards.
///
/// So there is no default here and deliberately no <c>Parse</c> that conjures one from a name. A
/// match carries a profile its players filled in, or it cannot be played. <see cref="Empty"/> is
/// blank rather than typical, and <see cref="Validate"/> says what is still missing.
/// </remarks>
/// <param name="Name">What the players call this set of numbers.</param>
/// <param name="DieFaces">Faces on the die the whole game is rolled with.</param>
/// <param name="BeamDamage">What a beam die scores at each face and screen level.</param>
/// <param name="BeamRangeBandWidth">Range band a beam loses a die per, in mu.</param>
/// <param name="MaxScreenLevel">Highest screen level a ship may carry.</param>
/// <param name="TorpedoMaximumRange">Longest reach of a pulse torpedo, in mu.</param>
/// <param name="TorpedoBandWidth">Range band a torpedo's to-hit number worsens per, in mu.</param>
/// <param name="TorpedoBestToHit">To-hit number in the closest band.</param>
/// <param name="NeedleBeamRange">Reach of a needle beam that declares no range of its own, in mu.</param>
/// <param name="NeedleSystemKillRoll">Roll that knocks the nominated system out.</param>
/// <param name="EnhancedNeedleBeams">True when a needle also puts a point into the hull.</param>
/// <param name="NeedleHullDamageRoll">Roll an enhanced needle needs to draw blood.</param>
/// <param name="ThresholdRows">Whether the damage track is a fixed number of rows or sized by band.</param>
/// <param name="ThresholdRowCount">Rows in the track when it is a fixed number.</param>
/// <param name="EscortRowCount">Rows an escort's track is drawn in, when sized by band.</param>
/// <param name="CruiserRowCount">Rows a cruiser's track is drawn in, when sized by band.</param>
/// <param name="MaxPartiesPerJob">Damage control parties that can usefully crowd one job.</param>
/// <param name="RepairRollWithOneParty">Roll one party needs to bring a system back.</param>
/// <param name="RepairBestRoll">Best roll any number of parties can bring the job down to.</param>
/// <param name="FighterMoveAllowance">How far a fighter group flies in a turn, in mu.</param>
/// <param name="CarrierRatesFollowBays">True when flight operations follow the ship's bays.</param>
/// <param name="TrueCarrierAllowance">Groups a true carrier works in a turn, when they do not.</param>
/// <param name="OtherShipAllowance">Groups any other ship with a bay works, when they do not.</param>
/// <param name="CarrierTurnaroundRoll">True when a landed group rolls before it may go out again.</param>
/// <param name="Turnaround">What each turnaround face means.</param>
/// <param name="PointDefenseRange">How far point defence reaches, in mu.</param>
/// <param name="PointDefenseKills">What each point-defence face shoots down.</param>
/// <param name="PointDefenseChainOnFace">Face that earns another die, or zero when none does.</param>
/// <param name="MissilesPerSalvo">Missiles in one salvo.</param>
/// <param name="SalvoAttackRadius">How close a target must be to the point of aim, in mu.</param>
public sealed record RulesProfile(
    string Name,
    int DieFaces,
    ImmutableArray<BeamDamageEntry> BeamDamage,
    int BeamRangeBandWidth,
    int MaxScreenLevel,
    int TorpedoMaximumRange,
    int TorpedoBandWidth,
    int TorpedoBestToHit,
    int NeedleBeamRange,
    int NeedleSystemKillRoll,
    bool EnhancedNeedleBeams,
    int NeedleHullDamageRoll,
    ThresholdRowMode ThresholdRows,
    int ThresholdRowCount,
    int EscortRowCount,
    int CruiserRowCount,
    int MaxPartiesPerJob,
    int RepairRollWithOneParty,
    int RepairBestRoll,
    int FighterMoveAllowance,
    bool CarrierRatesFollowBays,
    int TrueCarrierAllowance,
    int OtherShipAllowance,
    bool CarrierTurnaroundRoll,
    ImmutableArray<TurnaroundEntry> Turnaround,
    int PointDefenseRange,
    ImmutableArray<PointDefenseEntry> PointDefenseKills,
    int PointDefenseChainOnFace,
    int MissilesPerSalvo,
    int SalvoAttackRadius)
{
    /// <summary>
    /// A profile with nothing filled in. The starting point for a player entering their own numbers,
    /// and what an old saved match restores with until somebody supplies one.
    /// </summary>
    public static RulesProfile Empty { get; } = new(
        Name: string.Empty,
        DieFaces: 0,
        BeamDamage: [],
        BeamRangeBandWidth: 0,
        MaxScreenLevel: 0,
        TorpedoMaximumRange: 0,
        TorpedoBandWidth: 0,
        TorpedoBestToHit: 0,
        NeedleBeamRange: 0,
        NeedleSystemKillRoll: 0,
        EnhancedNeedleBeams: false,
        NeedleHullDamageRoll: 0,
        ThresholdRows: ThresholdRowMode.FixedRows,
        ThresholdRowCount: 0,
        EscortRowCount: 0,
        CruiserRowCount: 0,
        MaxPartiesPerJob: 0,
        RepairRollWithOneParty: 0,
        RepairBestRoll: 0,
        FighterMoveAllowance: 0,
        CarrierRatesFollowBays: false,
        TrueCarrierAllowance: 0,
        OtherShipAllowance: 0,
        CarrierTurnaroundRoll: false,
        Turnaround: [],
        PointDefenseRange: 0,
        PointDefenseKills: [],
        PointDefenseChainOnFace: 0,
        MissilesPerSalvo: 0,
        SalvoAttackRadius: 0);

    /// <summary>True when this profile has enough in it to play a match against.</summary>
    public bool IsPlayable => Validate().Count == 0;

    /// <summary>
    /// The same profile with every table guaranteed to be a real empty array rather than a default
    /// one.
    /// </summary>
    /// <remarks>
    /// A JSON body that leaves a table out deserialises it as <c>default(ImmutableArray&lt;T&gt;)</c>,
    /// which is not an empty array - it is a null-ish struct that throws the moment anything
    /// enumerates or serialises it. Sending a profile with no turnaround table therefore blew up on
    /// the way back out rather than on the way in, which is a miserable way to find out. Every
    /// profile crossing into the app goes through here.
    /// </remarks>
    public RulesProfile Normalized() => this with
    {
        Name = Name ?? string.Empty,
        BeamDamage = BeamDamage.IsDefault ? [] : BeamDamage,
        Turnaround = Turnaround.IsDefault ? [] : Turnaround,
        PointDefenseKills = PointDefenseKills.IsDefault ? [] : PointDefenseKills,
    };

    /// <summary>
    /// What is still missing before a match can be played against this profile.
    /// </summary>
    /// <returns>One line per gap, empty when the profile is complete.</returns>
    /// <remarks>
    /// Only the numbers every match needs are required. A table that has no salvo missiles in it
    /// does not have to describe them, so ordnance, needles and carrier operations are checked only
    /// once their own numbers are partly filled in - a half-entered section is a mistake, an empty
    /// one is a choice.
    /// </remarks>
    public IReadOnlyList<string> Validate()
    {
        var gaps = new List<string>();

        if (string.IsNullOrWhiteSpace(Name))
        {
            gaps.Add("The profile needs a name, so you can tell it from the next one.");
        }

        if (DieFaces < 2)
        {
            gaps.Add("Say how many faces the die has.");
        }

        if (BeamDamage.IsDefaultOrEmpty)
        {
            gaps.Add("Fill in what a beam die scores at each face and screen level.");
        }

        if (BeamRangeBandWidth <= 0)
        {
            gaps.Add("Say how wide a beam's range band is.");
        }

        if (ThresholdRowCount <= 1)
        {
            gaps.Add("Say how many rows a hull's damage track is drawn in.");
        }

        if (ThresholdRows == ThresholdRowMode.ByShipClass && (EscortRowCount <= 0 || CruiserRowCount <= 0))
        {
            gaps.Add("Sizing the track by class needs a row count for an escort and for a cruiser.");
        }

        if (MaxPartiesPerJob > 0 && (RepairRollWithOneParty <= 0 || RepairBestRoll <= 0))
        {
            gaps.Add("Damage control needs the roll one party makes and the best it can get to.");
        }

        if (PointDefenseRange > 0 && PointDefenseKills.IsDefaultOrEmpty)
        {
            gaps.Add("Point defence has a range but nothing saying what its dice shoot down.");
        }

        if (MissilesPerSalvo > 0 && SalvoAttackRadius <= 0)
        {
            gaps.Add("A salvo needs to know how close its target must be to the point of aim.");
        }

        if (EnhancedNeedleBeams && NeedleHullDamageRoll <= 0)
        {
            gaps.Add("An enhanced needle needs the roll it draws blood on.");
        }

        if (CarrierTurnaroundRoll && Turnaround.IsDefaultOrEmpty)
        {
            gaps.Add("Turnaround is rolled for but nothing says what the faces mean.");
        }

        // A table row for a face the die cannot show is a slip in the transcription, not a choice,
        // and one that fails quietly: the row is never matched, so the die scores nothing and the
        // table looks merely unlucky. Say so while the numbers are still on screen.
        if (DieFaces >= 2)
        {
            if (!BeamDamage.IsDefaultOrEmpty && BeamDamage.Any(entry => entry.DieFace < 1 || entry.DieFace > DieFaces))
            {
                gaps.Add($"A beam damage entry names a face the {DieFaces}-sided die does not have.");
            }

            if (!PointDefenseKills.IsDefaultOrEmpty && PointDefenseKills.Any(entry => entry.DieFace < 1 || entry.DieFace > DieFaces))
            {
                gaps.Add($"A point defence entry names a face the {DieFaces}-sided die does not have.");
            }

            if (!Turnaround.IsDefaultOrEmpty && Turnaround.Any(entry => entry.DieFace < 1 || entry.DieFace > DieFaces))
            {
                gaps.Add($"A turnaround entry names a face the {DieFaces}-sided die does not have.");
            }
        }

        return gaps;
    }

    /// <summary>
    /// What a beam die scores at this face against this much screening. A combination the profile
    /// does not mention scores nothing, so a player only has to enter the faces that do something.
    /// </summary>
    /// <param name="dieFace">The face rolled.</param>
    /// <param name="screenLevel">The target's screen level.</param>
    public int BeamDamageFor(int dieFace, int screenLevel)
    {
        var screens = Math.Clamp(screenLevel, 0, Math.Max(0, MaxScreenLevel));
        foreach (var entry in BeamDamage.IsDefault ? [] : BeamDamage)
        {
            if (entry.DieFace == dieFace && entry.ScreenLevel == screens)
            {
                return Math.Max(0, entry.Damage);
            }
        }

        return 0;
    }

    /// <summary>How many incoming one point-defence face accounts for.</summary>
    /// <param name="dieFace">The face rolled.</param>
    public int PointDefenseKillsFor(int dieFace)
    {
        foreach (var entry in PointDefenseKills.IsDefault ? [] : PointDefenseKills)
        {
            if (entry.DieFace == dieFace)
            {
                return Math.Max(0, entry.Kills);
            }
        }

        return 0;
    }

    /// <summary>
    /// What a turnaround face means. A face the profile does not mention leaves the group flyable
    /// and ready next turn, which is the least this roll can do to it.
    /// </summary>
    /// <param name="dieFace">The face rolled.</param>
    public TurnaroundEntry TurnaroundFor(int dieFace)
    {
        foreach (var entry in Turnaround.IsDefault ? [] : Turnaround)
        {
            if (entry.DieFace == dieFace)
            {
                return entry;
            }
        }

        return new TurnaroundEntry(dieFace, IsGroundedForGame: false, TurnsBeforeRelaunch: 1);
    }

    /// <summary>Compares two profiles by their contents, tables included.</summary>
    /// <param name="other">The profile to compare against.</param>
    public bool Equals(RulesProfile? other) =>
        other is not null
        && Name == other.Name
        && DieFaces == other.DieFaces
        && BeamRangeBandWidth == other.BeamRangeBandWidth
        && MaxScreenLevel == other.MaxScreenLevel
        && TorpedoMaximumRange == other.TorpedoMaximumRange
        && TorpedoBandWidth == other.TorpedoBandWidth
        && TorpedoBestToHit == other.TorpedoBestToHit
        && NeedleBeamRange == other.NeedleBeamRange
        && NeedleSystemKillRoll == other.NeedleSystemKillRoll
        && EnhancedNeedleBeams == other.EnhancedNeedleBeams
        && NeedleHullDamageRoll == other.NeedleHullDamageRoll
        && ThresholdRows == other.ThresholdRows
        && ThresholdRowCount == other.ThresholdRowCount
        && EscortRowCount == other.EscortRowCount
        && CruiserRowCount == other.CruiserRowCount
        && MaxPartiesPerJob == other.MaxPartiesPerJob
        && RepairRollWithOneParty == other.RepairRollWithOneParty
        && RepairBestRoll == other.RepairBestRoll
        && FighterMoveAllowance == other.FighterMoveAllowance
        && CarrierRatesFollowBays == other.CarrierRatesFollowBays
        && TrueCarrierAllowance == other.TrueCarrierAllowance
        && OtherShipAllowance == other.OtherShipAllowance
        && CarrierTurnaroundRoll == other.CarrierTurnaroundRoll
        && PointDefenseRange == other.PointDefenseRange
        && PointDefenseChainOnFace == other.PointDefenseChainOnFace
        && MissilesPerSalvo == other.MissilesPerSalvo
        && SalvoAttackRadius == other.SalvoAttackRadius
        && Same(BeamDamage, other.BeamDamage)
        && Same(Turnaround, other.Turnaround)
        && Same(PointDefenseKills, other.PointDefenseKills);

    /// <summary>Hashes a profile by its contents, tables included.</summary>
    public override int GetHashCode()
    {
        var hash = new HashCode();
        hash.Add(Name);
        hash.Add(DieFaces);
        hash.Add(BeamRangeBandWidth);
        hash.Add(MaxScreenLevel);
        hash.Add(ThresholdRowCount);
        hash.Add(MissilesPerSalvo);
        foreach (var entry in BeamDamage.IsDefault ? [] : BeamDamage)
        {
            hash.Add(entry);
        }

        return hash.ToHashCode();
    }

    /// <summary>
    /// Compares two tables element by element.
    /// </summary>
    /// <remarks>
    /// An <see cref="ImmutableArray{T}"/> compares by reference, so two profiles built from the same
    /// numbers would otherwise come out unequal - which has bitten this codebase before.
    /// </remarks>
    private static bool Same<T>(ImmutableArray<T> left, ImmutableArray<T> right) =>
        (left.IsDefaultOrEmpty && right.IsDefaultOrEmpty)
        || (!left.IsDefault && !right.IsDefault && left.SequenceEqual(right));
}

File: src/ForceSignal.Web/src/lib/rulesProfile.ts
import { blankRulesProfile, type RulesProfile } from '../types.ts';

const STORAGE_KEY = 'forcesignal.rulesProfiles';

/**
 * What is still missing before a profile can be played against.
 *
 * The server checks the same things and is the authority - this is here so the form can say what is
 * wrong before a round trip, not so the client can decide. If the two ever disagree, the server wins
 * and the request comes back refused.
 */
export function gapsIn(profile: RulesProfile): string[] {
  const gaps: string[] = [];
  if (!profile.name.trim()) {
    gaps.push('The profile needs a name, so you can tell it from the next one.');
  }
  if (profile.dieFaces < 2) {
    gaps.push('Say how many faces the die has.');
  }
  if (profile.beamDamage.length === 0) {
    gaps.push('Fill in what a beam die scores at each face and screen level.');
  }
  if (profile.beamRangeBandWidth <= 0) {
    gaps.push("Say how wide a beam's range band is.");
  }
  if (profile.thresholdRowCount <= 1) {
    gaps.push("Say how many rows a hull's damage track is drawn in.");
  }
  if (profile.thresholdRows === 'ByShipClass' && (profile.escortRowCount <= 0 || profile.cruiserRowCount <= 0)) {
    gaps.push('Sizing the track by class needs a row count for an escort and for a cruiser.');
  }
  if (profile.maxPartiesPerJob > 0 && (profile.repairRollWithOneParty <= 0 || profile.repairBestRoll <= 0)) {
    gaps.push('Damage control needs the roll one party makes and the best it can get to.');
  }
  if (profile.pointDefenseRange > 0 && profile.pointDefenseKills.length === 0) {
    gaps.push('Point defence has a range but nothing saying what its dice shoot down.');
  }
  if (profile.missilesPerSalvo > 0 && profile.salvoAttackRadius <= 0) {
    gaps.push('A salvo needs to know how close its target must be to the point of aim.');
  }
  if (profile.enhancedNeedleBeams && profile.needleHullDamageRoll <= 0) {
    gaps.push('An enhanced needle needs the roll it draws blood on.');
  }
  if (profile.carrierTurnaroundRoll && profile.turnaround.length === 0) {
    gaps.push('Turnaround is rolled for but nothing says what the faces mean.');
  }
  // A row for a face the die cannot show never matches, so the die scores nothing and the table
  // just looks unlucky. Same check as the server, said while the numbers are still on screen.
  if (profile.dieFaces >= 2) {
    const offTheDie = (entry: { dieFace: number }) => entry.dieFace < 1 || entry.dieFace > profile.dieFaces;
    if (profile.beamDamage.some(offTheDie)) {
      gaps.push(`A beam damage entry names a face the ${profile.dieFaces}-sided die does not have.`);
    }
    if (profile.pointDefenseKills.some(offTheDie)) {
      gaps.push(`A point defence entry names a face the ${profile.dieFaces}-sided die does not have.`);
    }
    if (profile.turnaround.some(offTheDie)) {
      gaps.push(`A turnaround entry names a face the ${profile.dieFaces}-sided die does not have.`);
    }
  }
  return gaps;
}

/**
 * Whether two profiles hold the same numbers.
 *
 * By content, not by identity, and that is the whole point of it: a profile arrives freshly parsed
 * inside every snapshot, so two objects that are not the same object are still almost always the
 * same numbers. Anything watching for the table's profile to change has to be able to tell those
 * two cases apart, or it fires on every ship anyone moves.
 *
 * Keys are sorted before comparing so that a profile built by the form and one parsed off the wire
 * compare equal despite having been assembled in different orders.
 */
export function sameProfile(left: RulesProfile, right: RulesProfile): boolean {
  return stableJson(left) === stableJson(right);
}

function stableJson(value: unknown): string {
  return JSON.stringify(value, (_key, item: unknown) => {
    if (item === null || typeof item !== 'object' || Array.isArray(item)) {
      return item;
    }
    const sorted: Record<string, unknown> = {};
    for (const key of Object.keys(item as Record<string, unknown>).sort()) {
      sorted[key] = (item as Record<string, unknown>)[key];
    }
    return sorted;
  });
}

/** What this browser has saved, and what it is holding that this version cannot give back. */
export type SavedProfiles = {
  /** The profiles, ready to load onto the form. */
  readonly profiles: readonly RulesProfile[];
  /** One sentence per stored entry this version cannot read whole. Empty in the ordinary case. */
  readonly problems: readonly string[];
};

/** The store exactly as it sits, entry by entry, with nothing read into it. */
function storedEntries(): unknown[] {
  try {
    const raw = window.localStorage.getItem(STORAGE_KEY);
    const parsed: unknown = raw ? JSON.parse(raw) : [];
    return Array.isArray(parsed) ? parsed : [];
  } catch {
    // A store whose text will not parse at all holds nothing this app can hand back entry by entry,
    // and there is no honest per-entry report to make about it either. The next save does overwrite
    // it; refusing to save instead would strand the player with no way out of a state only they can
    // see. What is preserved is the case that can be: entries inside a store that does parse.
    return [];
  }
}

/** The name a stored entry was saved under, for saying which one is being talked about. */
function storedName(entry: unknown): string | null {
  if (typeof entry !== 'object' || entry === null || Array.isArray(entry)) {
    return null;
  }

  const name = (entry as { name?: unknown }).name;
  return typeof name === 'string' && name.trim() ? name : null;
}

/**
 * Profiles this browser has saved, so a set of numbers is entered once and reused.
 *
 * This used to be `parsed.map(readProfile)`, and `readProfile` coerces - that is its job, because a
 * profile written by an older version is still a profile. Asked to read something that is *not* one
 * it answers with a profile full of zeros, so anything that ended up under this key came back
 * looking like a saved set of numbers, offered in the picker, and blanked the form the moment it
 * was chosen. That is the import bug through the other door, and this door had nowhere to report
 * into, so it happened in silence.
 *
 * Two answers, and which one an entry gets is the whole of it:
 *
 * - **Not a profile at all** - it fails the same `looksLikeProfile` bar the import uses. Not
 *   offered, and named in `problems`. It is left in the store untouched.
 * - **A profile carrying field names this version does not have** - a profile saved by a build that
 *   called two of its fields something else. Still offered, because it is the player's and most of
 *   it reads, with the unreadable part named so the blanks on the form are explained rather than
 *   discovered.
 *
 * What is *not* refused is incompleteness. A profile does not need all thirty fields to be usable -
 * the editor's own hints say to leave the torpedo reach and the salvo size at zero - so a stored
 * profile that omits them is a good profile and is handed back without comment.
 */
export function savedProfiles(): SavedProfiles {
  const profiles: RulesProfile[] = [];
  const problems: string[] = [];

  for (const entry of storedEntries()) {
    const named = storedName(entry);
    if (!looksLikeProfile(entry)) {
      problems.push(
        `Something saved in this browser${named ? ` under "${named}"` : ''} could not be read as a `
        + 'rules profile, so it is not offered above. It has been left where it is.',
      );
      continue;
    }

    profiles.push(readProfile(entry));

    const unknown = fieldsNotReadFrom(entry);
    if (unknown.length > 0) {
      problems.push(
        `"${named ?? profiles[profiles.length - 1].name}" was saved with ${unknown.length} field`
        + `${unknown.length === 1 ? '' : 's'} this version does not read, so ${unknown.length === 1
          ? 'it will not'
          : 'they will not'} appear on the form: ${unknown.join(', ')}.`,
      );
    }
  }

  return { profiles, problems };
}

/**
 * Field names a stored entry carries that this version of `RulesProfile` has no place for.
 *
 * The mirror of `fieldsBlankedBy`: that one asks what a file would cost the form, this one asks what
 * the form cannot show of a file. A renamed field is carried through a read and a write untouched -
 * `readProfile` spreads the source over the blank - but nothing ever displays it, so the number the
 * player typed is on disk and off the screen, and the field that replaced it reads zero.
 */
function fieldsNotReadFrom(value: unknown): string[] {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    return [];
  }

  const fields = new Set(Object.keys(blankRulesProfile));
  return Object.keys(value).filter((key) => !fields.has(key));
}

/**
 * Saves a profile under its own name, replacing any earlier one by that name.
 *
 * Written back off the *stored* entries rather than off the ones that read cleanly. That is not
 * tidiness: once the reader above stopped handing back a coerced version of an entry it cannot
 * read, a writer built on the reader would have dropped that entry from the store on the next save,
 * and the repair for a silent blanking would have become a silent deletion.
 */
export function saveProfile(profile: RulesProfile): SavedProfiles {
  const kept = storedEntries().filter((entry) => storedName(entry) !== profile.name);
  window.localStorage.setItem(STORAGE_KEY, JSON.stringify(sortedByName([...kept, profile])));
  return savedProfiles();
}

/** Forgets a saved profile by name, and only that one. */
export function deleteProfile(name: string): SavedProfiles {
  const kept = storedEntries().filter((entry) => storedName(entry) !== name);
  window.localStorage.setItem(STORAGE_KEY, JSON.stringify(sortedByName(kept)));
  return savedProfiles();
}

function sortedByName(entries: unknown[]): unknown[] {
  return [...entries].sort((left, right) => (storedName(left) ?? '').localeCompare(storedName(right) ?? ''));
}

/**
 * Reads a profile out of parsed JSON, filling anything absent from the blank profile.
 *
 * Unknown fields are dropped and missing ones default to blank rather than to a number of this
 * app's choosing, so a half-written file comes back as a half-filled form the player can finish
 * instead of a profile that quietly plays wrong.
 */
function readProfile(value: unknown): RulesProfile {
  const source = (value ?? {}) as Partial<RulesProfile>;
  return {
    ...blankRulesProfile,
    ...source,
    name: typeof source.name === 'string' ? source.name : '',
    beamDamage: Array.isArray(source.beamDamage) ? source.beamDamage : [],
    turnaround: Array.isArray(source.turnaround) ? source.turnaround : [],
    pointDefenseKills: Array.isArray(source.pointDefenseKills) ? source.pointDefenseKills : [],
    thresholdRows: source.thresholdRows === 'ByShipClass' ? 'ByShipClass' : 'FixedRows',
  };
}

/**
 * Whether a parsed file is a rules profile at all, rather than some other JSON.
 *
 * `readProfile` coerces - it has to, because a profile written by an older version is still a
 * profile - and a coercer asked to read a fleet export answers with a profile full of zeros. So the
 * question "is this one of ours?" is asked before it, and the answer is read off the file's own
 * field names: two or more of them means a profile someone wrote, and nothing else in this app
 * writes a file that clears that bar.
 */
function looksLikeProfile(value: unknown): boolean {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    return false;
  }

  const fields = new Set(Object.keys(blankRulesProfile));
  return Object.keys(value).filter((key) => fields.has(key)).length >= 2;
}

/**
 * The fields a parsed file would blank, given what is on the form now.
 *
 * Not the same question as "what is the file missing?", and the difference is the whole of it. A
 * profile does not need all thirty fields to be usable - the editor's own hints say to leave the
 * torpedo reach and the salvo size at zero if your table does not use them - so a file that omits
 * them is a perfectly good file, and refusing it would be this app deciding which optional rules a
 * table has to play. What is never acceptable is the *loss*: a field the player has filled in going
 * to zero because the file said nothing about it.
 *
 * So the question asked is the one that matters. A field the file does not carry costs nothing when
 * the form has nothing in it either, and costs the player their own number when it does.
 *
 * Walked off `blankRulesProfile` rather than off a list kept here, so a field added to
 * `RulesProfile` tomorrow is covered the day it is added.
 */
function fieldsBlankedBy(value: unknown, current: RulesProfile): string[] {
  const carried = typeof value === 'object' && value !== null && !Array.isArray(value)
    ? new Set(Object.keys(value))
    : new Set<string>();

  const blank = blankRulesProfile as unknown as Record<string, unknown>;
  const filled = current as unknown as Record<string, unknown>;
  return Object.keys(blank).filter((field) =>
    !carried.has(field) && JSON.stringify(filled[field]) !== JSON.stringify(blank[field]));
}

/** What an import that could not be read at all says, wherever it failed. */
export const unreadableProfileMessage =
  'That file could not be read as a rules profile, so the numbers on screen have been left as they '
  + 'are. Pick the file you exported from here, or type the numbers in.';

/** Either a whole profile read off a file, or the reason the file was not applied. */
export type ProfileImport =
  | { readonly ok: true; readonly profile: RulesProfile }
  | { readonly ok: false; readonly problem: string };

/**
 * Reads the bytes of a picked file, and refuses anything that would cost the player numbers.
 *
 * Two doors, and both used to be open. The first - a file that is not a profile at all - was closed
 * by `looksLikeProfile`, which asks whether the file carries two or more of these field names. The
 * second is the one that bar cannot see: **a file that is unmistakably one of ours and is missing
 * most of it.** `{ name, dieFaces }` clears a two-field bar, and spreading it over the blank profile
 * wrote zeros into the other twenty-eight fields the player had typed off their own rulebook - a
 * truncated export, a half-written file, or a profile from a version that renamed the rest, and no
 * message either way.
 *
 * What is refused is the **loss**, not the incompleteness. Those are two different things and
 * getting them confused costs a real capability: a profile does not need all thirty fields to be
 * usable, and a file that leaves out the torpedo and salvo numbers because that table does not play
 * them is a good file. Refusing it would be this app deciding which optional rules a table has to
 * use, which is the same sin in a different coat. So a partial file lands on an empty form, and is
 * turned away only when it would blank something already on the form - by name, all of them,
 * because the player has to know which of their numbers was at stake.
 *
 * @param current What is on the form now, and therefore what there is to lose.
 */
export function readProfileFile(text: string, current: RulesProfile = blankRulesProfile): ProfileImport {
  let payload: unknown;
  try {
    payload = JSON.parse(text);
  } catch {
    return { ok: false, problem: unreadableProfileMessage };
  }

  if (!looksLikeProfile(payload)) {
    return { ok: false, problem: unreadableProfileMessage };
  }

  const blanked = fieldsBlankedBy(payload, current);
  if (blanked.length > 0) {
    return {
      ok: false,
      problem: `That file is a rules profile, but it says nothing about ${blanked.length} field`
        + `${blanked.length === 1 ? '' : 's'} you have already filled in, so it would blank `
        + `${blanked.length === 1 ? 'it' : 'them'}. Nothing on screen has been changed. Start Blank `
        + `first if you meant to replace the lot: ${blanked.join(', ')}.`,
    };
  }

  return { ok: true, profile: readProfile(payload) };
}

/** The profile as a file a player can keep or pass to the rest of the table. */
export function exportProfile(profile: RulesProfile): void {
  const blob = new Blob([JSON.stringify(profile, null, 2)], { type: 'application/json' });
  const url = URL.createObjectURL(blob);
  const link = document.createElement('a');
  link.href = url;
  link.download = `${profile.name.trim().replace(/[^\w-]+/g, '-') || 'rules-profile'}.json`;
  link.click();
  URL.revokeObjectURL(url);
}
